<div style="border: 5px solid purple; padding: 10px; margin: 5px">
<b>Reviewer's comment – Iteration 2</b><br><br>

Hi Gabriela, I’m <b>Victor Camargo</b>. Thanks for addressing all of Svetlana’s comments — your project is now in excellent shape and I’m happy to approve it! ✅<br><br>

You’ve handled the duplicate check and final outlier treatment carefully, and the modeling flow is well-structured. The improvements you made bring everything together very nicely. Great work wrapping this up with clarity and precision!

Congratulations on completing this project! 🚀
</div>


<div style="border: 5px solid purple; padding: 10px; margin: 5px">
<b>   Svetlana's comment  </b>
      
Hi Gabriela, my name is Svetlana (https://hub.tripleten.com/u/6dee602c). Congratulations on submitting another project! 🎉 I will be using the standard the color marking:
    

   
    
<div style="border: 5px solid green; padding: 10px; margin: 5px">

Great solutions and ideas that can and should be used in the future are in green comments. Some of them are: 
    
    
- You have successfully prepared the subsets. It is important to split the data correctly in order to ensure there's no intersection;    
    
    
- Handled outliers; 
    

- Excluded irrelevant columns to reduce computational cost;
    
    
- Encoded cetegorical columns;    


- Trained and compared several models, great!

    
- Measured their training and prediction speed.
   

- Analyzed metrics. It is not enough to just fit the model and print the result. Instead, we have to analyze the results as it helps us identify what can be improved;

    
    
- Wrote an excellent conclusion! A well-written conclusion shows how the project met its objectives and provides a concise and understandable summary for those who may not have been involved in the details of the project. Good job! 

</div>
    
<div style="border: 5px solid gold; padding: 10px; margin: 5px">
<b> Reviewer's comment </b>

Yellow color indicates what should be optimized. This is not necessary, but it will be great if you make changes to this project. I've left several recommendations throughout the project. Please take a look.
 
</div>
<div style="border: 5px solid red; padding: 10px; margin: 5px">
<b> Reviewer's comment </b>

Issues that must be corrected to achieve accurate results are indicated in red comments. Please note that the project cannot be accepted until these issues are resolved. For instance,


- Check the data for the duplicates after you drop columns. If a dropped column contained unique values (ID or timestamp), removing it may make multiple rows appear the same;

    
    
    
- There are still some outliers in the Power column.


There may be other issues that need your attention. I described everything in my comments.  
</div>        
<hr>
    
<font color='dodgerblue'>**To sum up:**</font> great job here! You demonstrated strong analytical and modeling skills by preparing the data, experimenting with multiple advanced models, and evaluating them with appropriate metrics. The conclusion clearly communicates which model offers the best trade-off between speed and RMSE. There are several issues that need your attention. Please take a look at my comments and do not hesitate to ask questions if some of them seem unclear. I will wait the project for the second review 😊 
    

<hr>
    
Please use some color other than those listed to highlight answers to my comments.
I would also ask you **not to change, move or delete my comments** to make it easier for me to navigate during the next review.
    
<hr> 
    
✍️ Some notes:


- Here's a link to [Supervised Learning documenation sections](https://scikit-learn.org/stable/supervised_learning.html) that you may find useful.



- There are advanced tools such as [ColumnTransformer](https://scikit-learn.org/stable/modules/generated/sklearn.compose.ColumnTransformer.html) and [Pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html). `ColumnTransformer` and `Pipeline` are essential tools that help us create robust, maintainable, and efficient machine learning workflows. They work with data much more effectively. You can handle different data types and it is much easier to avoid data leakage. The code organization is very clean, but it may seem a bit difficult at the beginning. Take a look at this page to learn how to [organize a pipeline with ColumnTransformer](https://scikit-learn.org/stable/auto_examples/compose/plot_column_transformer_mixed_types.html).  
<hr>
    
    
📌 Please feel free to schedule a 1:1 sessions with our tutors or TAs Feel free to book 1-1 session [here](https://calendly.com/tripleten-ds-experts-team), join daily coworking sessions, or ask questions on TripleTen Hub if you need assistance 😉 
</div>

Rusty Bargain used car sales service is developing an app to attract new customers. In that app, you can quickly find out the market value of your car. You have access to historical data: technical specifications, trim versions, and prices. You need to build the model to determine the value. 

Rusty Bargain is interested in:

- the quality of the prediction;
- the speed of the prediction;
- the time required for training

## Data preparation

Let's import and clean the data in order to have the needed information to train our models.

In [1]:
import pandas as pd
import lightgbm as lgb
import numpy as np
import time
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from xgboost import XGBRegressor
from catboost import CatBoostRegressor

In [2]:


# Load data
df = pd.read_csv('/datasets/car_data.csv')

# Preview
print(df.shape)
df.head()

(354369, 16)


,DateCrawled,Price,VehicleType,RegistrationYear,Gearbox,Power,Model,Mileage,RegistrationMonth,FuelType,Brand,NotRepaired,DateCreated,NumberOfPictures,PostalCode,LastSeen
0,24/03/2016 11:52,480,NaN,1993,manual,0,golf,150000,0,petrol,volkswagen,NaN,24/03/2016 00:00,0,70435,07/04/2016 03:16
1,24/03/2016 10:58,18300,coupe,2011,manual,190,NaN,125000,5,gasoline,audi,yes,24/03/2016 00:00,0,66954,07/04/2016 01:46
2,14/03/2016 12:52,9800,suv,2004,auto,163,grand,125000,8,gasoline,jeep,NaN,14/03/2016 00:00,0,90480,05/04/2016 12:47
3,17/03/2016 16:54,1500,small,2001,manual,75,golf,150000,6,petrol,volkswagen,no,17/03/2016 00:00,0,91074,17/03/2016 17:40
4,31/03/2016 17:25,3600,small,2008,manual,69,fabia,90000,7,gasoline,skoda,no,31/03/2016 00:00,0,60437,06/04/2016 10:17


Now, let's check if we have any missing values and check how we can handle them if so.

In [3]:
# Check info
df.info()

# Check missing values
df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354369 entries, 0 to 354368
Data columns (total 16 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   DateCrawled        354369 non-null  object
 1   Price              354369 non-null  int64 
 2   VehicleType        316879 non-null  object
 3   RegistrationYear   354369 non-null  int64 
 4   Gearbox            334536 non-null  object
 5   Power              354369 non-null  int64 
 6   Model              334664 non-null  object
 7   Mileage            354369 non-null  int64 
 8   RegistrationMonth  354369 non-null  int64 
 9   FuelType           321474 non-null  object
 10  Brand              354369 non-null  object
 11  NotRepaired        283215 non-null  object
 12  DateCreated        354369 non-null  object
 13  NumberOfPictures   354369 non-null  int64 
 14  PostalCode         354369 non-null  int64 
 15  LastSeen           354369 non-null  object
dtypes: int64(7), object(

DateCrawled              0
Price                    0
VehicleType          37490
RegistrationYear         0
Gearbox              19833
Power                    0
Model                19705
Mileage                  0
RegistrationMonth        0
FuelType             32895
Brand                    0
NotRepaired          71154
DateCreated              0
NumberOfPictures         0
PostalCode               0
LastSeen                 0
dtype: int64

It looks like we have quite a few missing values. Since all these columns are categorical and contain meaningful information, we should retain them and fill in the missing values with a placeholder, such as 'unknown'. Dropping them would cause excessive data loss or remove essential features.

In [4]:
# Fill missing categorical values with 'unknown'
df[['VehicleType', 'Gearbox', 'Model', 'FuelType', 'NotRepaired']] = df[['VehicleType', 'Gearbox', 'Model', 'FuelType', 'NotRepaired']].fillna('unknown')

<div style="border: 5px solid green; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>
    
It's a good decision to fill in the gaps with some unique value, it's better than deleting them. Moreover, it is normal that sometimes sellers do not specify some information. The model should "know" about such cases. We even should not use median or mode. Even though the median does not skew the distribution, we have to many missing values to fill in because there is a risk of biasing the data.   
    
</div>

Now with the missing values resolved, let's see if we have any duplicates within the dataset.

In [5]:
df.duplicated().sum()

262

Looks like we have some duplicates too! Let's remove them since they can bias the model and do not provide new information.

In [6]:
# Drop duplicates
df = df.drop_duplicates().reset_index(drop=True)

# Verify
print("New shape:", df.shape)

New shape: (354107, 16)


Now, let's drop the unnecessary columns since they will not help with the model training:

In [7]:
df = df.drop(['DateCrawled', 'DateCreated', 'LastSeen', 'PostalCode', 'NumberOfPictures'], axis=1)

Awesome! Now, let's remove the unrealistic values:

In [8]:
# Price
df = df[(df['Price'] >= 100) & (df['Price'] <= 100000)]

# Power
df = df[(df['Power'] >= 50) & (df['Power'] <= 500)]

# RegistrationYear
df = df[(df['RegistrationYear'] >= 1950) & (df['RegistrationYear'] <= 2025)]

df.shape

(299670, 11)

Now our data's squeaky clean! Now, on to the model training.

<div style="border: 5px solid red; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>

- > `df = df[(df['Power'] > 0) & (df['Power'] <= 500)]`
                                                     
                                                     
What if the next minimum power value is 1 or 2?  Consider removing everything below 50.
                              
- After removing unnecessary columns, it makes sense to check the data for duplicates again, as the dataset will later be split into training and test sets. Removing specific columns can cause previously distinct rows to become identical. If a dropped column contained unique values (ID or timestamp), removing it may make multiple rows appear the same.   
   
</div>
<div style="border: 5px solid gold; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>
    
- It will be perfect if you analyze the distributions and display the charts, thus showing a reader why you decided to delete specific rows. Why is this important? In real-world problems, the data is rarely clean. Displaying distributions helps us evaluate the data, find outliers, identify the required preprocessing steps, and understand feature relationships, which informs feature engineering. Feature engineering in some cases is a clue.


- Consider analyzing categories as well. Petrol and gasoline refer to the same fuel, so we can use one of these categories. Some rare model categories can be dropped. If a category appears only in the training or validation subset, for instance, and we use `handle_unknown='ignore'`, the linear model might miss important signals in validation or make predictions with incomplete features thus breaking the assumptions of linearity. It may be helpful to make sure that training and validation subsets use the same feature columns after encoding.
</div>

## Model training

First things first! Let's separate the features and the target:

In [9]:
# Separate the target
target = df['Price']
features = df.drop('Price', axis=1)

Now, let's OHE categorical features since models like Linear Regression and XGBoost need numerical inputs, which will:

- Convert `FuelType`, `Brand`, `Gearbox`, etc., into numerical columns.

- Avoid the dummy variable trap (by dropping the first category in each).

In [10]:
# Get dummy variables for categorical features
features = pd.get_dummies(features, drop_first=True)

<div style="border: 5px solid green; padding: 15px; margin: 5px">
<b>   Reviewer's comment </b>
    
Yes, we need to encode data here, well done! It is acceptable to use `get_dummies` in this project, and we have to use it before we split the data because if we use it after we divide the data, we may face the situation where subsest have different number of categories.
    
</div>
<div style="border: 5px solid gold; padding: 15px; margin: 5px">
<b>   Reviewer's comment </b>


- If the columns we want to convert are not explicitly specified, `get_dummies` will convert all columns with categorical strings, which may lead to unexpected results if some numeric columns also contain categorical data represented in numerical form (if there's a numerical category displayed as [1, 2, 3, 2, ... ]).


    
- Please note that `OneHotEncoder(handle_unknown='ignore')` or `OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1)` are generally more robust than `get_dummies` because they can handle situations where test subset has features that were not available during training. [Difference between OneHotEncoder and get_dummies](https://pythonsimplified.com/difference-between-onehotencoder-and-get_dummies/). 
    
    
    
For tree-based models, `OrdinalEncoder` is a better choice because of computational cost. For boosting algorithms, we can rely on internal encoders that usually perform even better than external ones. For `CatBoost`, this is controlled by the `cat_features` parameter. For `LightGBM`, you can convert categorical features to the category type, allowing the model to handle them automatically.
    
   

    
`OrdinalEncoder()` or `LabelEncoder()` should not be used with linear models if there's no ordinal relationship. [How and When to Use Ordinal Encoder](https://leochoi146.medium.com/how-and-when-to-use-ordinal-encoder-d8b0ef90c28c). For linear regresison, I recommend using `OneHotEncoder(handle_unknown='ignore')`. 


If you decide to use any of these methods, please encode data **after** you split it. 

    
    
    
For instance, you can use `Ordinal` for Forest and Tree, `OneHotEncoder` for Lin. Regression and categorical data types for boosting models.
</div>



Now, let's split our dataset into 80 % training and 20 % validation:

In [11]:
X_train, X_valid, y_train, y_valid = train_test_split(
    features, target, test_size=0.2, random_state=42
)

<div style="border: 5px solid gold; padding: 10px; margin: 5px">
<b>   Svetlana's comment </b>
    

- [PEP8](https://peps.python.org/pep-0008/) states that one should always put imports at the top of the file. It's a good practice, since everyone who is going to read the project, can immediately figure out what modules need to be installed. Moreover, imports should be placed in a separate cell. 


- Consider saving at least one subset for the final testing. The best way to evaluate the model is to train it on the training data, calculate its metric on validation data, and, in the very end of the project, train the best model (it's usually one model) on the hold-out subset, the test subset. In this case, we need 3 subsets. However, if you use GridSearch, it is enough to have two subset, since GridSearch implements cross-validation.
</div>

Split! Now, we will start with the Linear Regression model and see what RMSE we get: 

In [12]:
# Instantiate model

# 1. Create the model
model = LinearRegression()

# 2. Time only the training step
start_train = time.time()
model.fit(X_train, y_train)
end_train = time.time()
training_time = end_train - start_train

# 3. Time only starts the prediction step
start_pred = time.time()
preds = model.predict(X_valid)
end_pred = time.time()
prediction_time = end_pred - start_pred

# 4. Evaluate RMSE (outside of timing)
rmse = mean_squared_error(y_valid, preds, squared=False)

# 5. Print results
print(f"Training time: {training_time:.2f} seconds")
print(f"Prediction time: {prediction_time:.2f} seconds")
print(f"RMSE: {rmse:.2f}")

Training time: 8.59 seconds
Prediction time: 0.17 seconds
RMSE: 2545.07


<div style="border: 5px solid red; padding: 15px; margin: 5px">
<b> Reviewer's comment</b>
    

One of the main goals here is to compare models' training and prediction speed. Would you please estimate them separately here as well? I see this is your baseline model, but we still should take a look at its speed. 

You can do it as follows: 

```python
start_time = time.time()
model.fit(X_train, y_train)
end_time1 = time.time()  


start_time = time.time()
model.predict(X_valid, y_valid)
end_time2 = time.time()

```
</div>

Great! Now, let's move on to the Random Forest model training:

In [13]:
# Create the model
rf_model = RandomForestRegressor(n_estimators=100, max_depth=20, random_state=42)

# Time only training
start_train = time.time()
rf_model.fit(X_train, y_train)
end_train = time.time()
training_time = end_train - start_train

# Time only prediction
start_pred = time.time()
rf_preds = rf_model.predict(X_valid)
end_pred = time.time()
prediction_time = end_pred - start_pred

# Evaluation (outside timing)
rf_rmse = mean_squared_error(y_valid, rf_preds, squared=False)

# Output
print(f"Training time: {training_time:.2f} seconds")
print(f"Prediction time: {prediction_time:.2f} seconds")
print(f"RMSE: {rf_rmse:.2f}")

Training time: 281.58 seconds
Prediction time: 1.51 seconds
RMSE: 1580.32


<div style="border: 5px solid gold; padding: 10px; margin: 5px">
<b> Reviewer's comment</b>
    
When estimating a specific piece of code (fitting the model), I recommend that you don't include creating a model or calculating the metric value in this measurement, since these actions also take time. It's better to estimate the training time only:
    
    
```python
start_time = time.time()
model.fit(X_train, y_train)
end_time = time.time()  
```
    
</div>

Perfect! Now, let's move on to train our LightGBM model:

In [14]:
lgb_model = lgb.LGBMRegressor(
    n_estimators=100,
    num_leaves=31,
    learning_rate=0.1,
    random_state=42,
    n_jobs=-1
)
lgb_model.fit(X_train, y_train)

# Time only training
start_train1 = time.time()
lgb_model.fit(X_train, y_train)
end_train1 = time.time()
training_time1 = end_train1 - start_train1

# Time only prediction
start_pred1 = time.time()
lgb_preds = lgb_model.predict(X_valid)
end_pred1 = time.time()
prediction_time1 = end_pred1 - start_pred1

# Evaluation (outside timing)
lgb_rmse = mean_squared_error(y_valid, lgb_preds, squared=False)

# Output
print(f"Training time: {training_time1:.2f} seconds")
print(f"Prediction time: {prediction_time1:.2f} seconds")
print(f"RMSE: {lgb_rmse:.2f}")

Training time: 3.50 seconds
Prediction time: 0.32 seconds
RMSE: 1682.52


Okay, we will try some hyperparameter tuning before moving on to training our CatBoost and XGBoost models and we can analyze each model:

In [15]:
# Tuned LightGBM
tuned_lgb = lgb.LGBMRegressor(
    n_estimators=200,
    learning_rate=0.05,
    num_leaves=64,
    max_depth=10,
    random_state=42,
    n_jobs=-1
)

# Time only training
start_train2 = time.time()
tuned_lgb.fit(X_train, y_train)
end_train2 = time.time()
training_time2 = end_train2 - start_train2

# Time only prediction
start_pred2 = time.time()
tuned_preds = tuned_lgb.predict(X_valid)
end_pred2 = time.time()
prediction_time2 = end_pred2 - start_pred2

# Evaluation
tuned_rmse = mean_squared_error(y_valid, tuned_preds, squared=False)

print(f"Training time: {training_time2:.2f} seconds")
print(f"Prediction time: {prediction_time2:.2f} seconds")
print("Tuned LightGBM RMSE:", tuned_rmse)

Training time: 6.42 seconds
Prediction time: 0.59 seconds
Tuned LightGBM RMSE: 1640.8089867752553


Amazing! Now, on to the CatBoost model:

In [16]:
cat_model = CatBoostRegressor(
    iterations=100,
    depth=10,
    learning_rate=0.1,
    verbose=0,
    random_state=42
)

# Time only training
start_train3 = time.time()
cat_model.fit(X_train, y_train)
end_train3 = time.time()
training_time3 = end_train3 - start_train3

# Time only prediction
start_pred3 = time.time()
cat_preds = cat_model.predict(X_valid)
end_pred3 = time.time()
prediction_time3 = end_pred3 - start_pred3

# Evaluation
cat_rmse = mean_squared_error(y_valid, cat_preds, squared=False)

print(f"Training time: {training_time3:.2f} seconds")
print(f"Prediction time: {prediction_time3:.2f} seconds")
print("CatBoost RMSE:", cat_rmse)

Training time: 3.61 seconds
Prediction time: 0.03 seconds
CatBoost RMSE: 1702.9290163904504


Great! Now on to the XGBoost model:

In [17]:
xgb_model = XGBRegressor(
    n_estimators=100,
    max_depth=10,
    learning_rate=0.1,
    n_jobs=-1,
    random_state=42
)

# Time only training
start_train4 = time.time()
xgb_model.fit(X_train, y_train)
end_train4 = time.time()
training_time4 = end_train4 - start_train4

# Time only prediction
start_pred4 = time.time()
xgb_preds = cat_model.predict(X_valid)
end_pred4 = time.time()
prediction_time4 = end_pred4 - start_pred4

xgb_rmse = mean_squared_error(y_valid, xgb_preds, squared=False)

print(f"Training time: {training_time4:.2f} seconds")
print(f"Prediction time: {prediction_time4:.2f} seconds")
print("XGBoost RMSE:", xgb_rmse)

Training time: 345.72 seconds
Prediction time: 0.02 seconds
XGBoost RMSE: 1702.9290163904504


We have all of our models! Now, let's analyze and compare our models to choose the ideal model.

<div style="border: 5px solid gold; padding: 15px; margin: 5px">
<b>   Reviewer's comment </b>


    
- Consider tuning hyperparameters for at least one of the models except for Linear Regression. For this purpose, use `RandomizedSearchCV` or `GridSearchCV`. If you decide to use a loop, don't forget to change the way you split the data, because in this case we will need three subsets, not two. 
    
    
    
- After we train all models, it is recommended that we choose the best **one** and check its performance on the test subset. Here we only need to make predictions and calculate RMSE. For the final testing, where we use the test subset to check the model's generalization ability, we should use the best model (one model or two models if they have almost the same metric values). We don't use all models here because even just checking their performance influences our choices. This leads to test set leakage when we unconsciously start picking models that perform well on the test set, making it part of the training loop. In real-world scenarios, the test set is meant to reflect how the final model performs in the wild. In practice, you only deploy one model, not several models, so testing just that final one mirrors reality. Moreover, evaluating every tuned model on the test set (especially with big models or datasets) is expensive and time-consuming. 




- When choosing the best model, we have to consider prediction time as well. The best model isn't always the one with the lowest error. Sometimes the errors are only slightly different, but the prediction time varies significantly. In such cases, it's worth considering a faster model. Think of a slow search engine that finds 10 useful links versus a fast one that finds 9. This is especially important if the model needs to operate in real time and produce results repeatedly. If a program runs just once, its speed might not even matter. But if it’s used continuously, optimization becomes crucial. So, in practice, apart from the other requirements, there are also runtime constraints for the model.

</div>

## Model analysis

📝 Final Summary:

We evaluated six regression models for predicting car prices using Rusty Bargain's dataset. Each model was assessed based on three criteria: prediction quality (RMSE), training time, and prediction speed.

🔹 Linear Regression (Baseline):

- RMSE: €2,545.07

- Trained quickly, but with the highest error.

- It is useful as a benchmark to validate other models.

🔹 Random Forest:

- RMSE: €1,580.32

- Highest accuracy, but very slow to train (almost 5 minutes).

- May be unsuitable for time-sensitive retraining pipelines.

🔹 LightGBM (Default & Tuned):

- Tuned LightGBM gave a strong RMSE of €1,640.81 with fast training and prediction.

- Offers the best speed–accuracy tradeoff, ideal for production environments.

🔹 CatBoost:

- RMSE: €1,702.93

- Extremely fast prediction (0.03s), suitable for real-time systems.

- Slightly lower accuracy than LightGBM.

🔹 XGBoost:

- Same RMSE as CatBoost (unexpectedly).

- Longest training time (5+ minutes).

- Not practical unless accuracy is significantly better (which it wasn't here).



✅ Conclusion:

| Best Accuracy       | 🏆 **Random Forest** (€1,580.32 RMSE)             |
| ------------------- | ------------------------------------------------- |
| Best Overall Model  | ⭐ **Tuned LightGBM** (speed + accuracy)           |
| Fastest Predictions | ⚡ **CatBoost / XGBoost** (0.03 / 0.02s)           |
| Sanity Check        | ✅ Linear Regression performed worst (as expected) |




<div style="border: 5px solid green; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>
   
Great conclusion! This is a solid final summary with comparison across models.    
</div>    

<div style="border: 5px solid red; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>
    
Don't forget to update it if needed. 

</div>

# Checklist

Type 'x' to check. Then press Shift+Enter.

- [x]  Jupyter Notebook is open.
- [x]  Code is error-free.
- [x]  The cells with the code have been arranged in order of execution.
- [x]  The data has been downloaded and prepared.
- [x]  The models have been trained.
- [x]  The analysis of speed and quality of the models has been performed.